File: backend/jasprchain/crypto/hash.py
"""Cryptographic Hash Functions for JasprChain
Mirrors Rust: sha2, blake3 crates
"""
import hashlib
from typing import List, Any
import json


def sha256_hash(data: bytes) -> bytes:
    """SHA-256 hash - used for transactions and addresses"""
    return hashlib.sha256(data).digest()


def sha256_hex(data) -> str:
    """SHA-256 hash as hex string"""
    if isinstance(data, str):
        data = data.encode('utf-8')
    return hashlib.sha256(data).hexdigest()


def double_sha256(data: bytes) -> bytes:
    """Double SHA-256 - used for block hashes"""
    return sha256_hash(sha256_hash(data))


def blake3_hash(data: bytes) -> bytes:
    """BLAKE3 hash - faster alternative for state hashing
    
    In production Rust: use blake3 crate
    Here we simulate with SHA-256 for compatibility
    """
    # BLAKE3 would be: blake3.blake3(data).digest()
    return hashlib.sha256(data).digest()


def merkle_root(items: List[bytes]) -> bytes:
    """Calculate Merkle root from list of hashes
    
    Used for:
    - Transaction root in blocks
    - State root in Sparse Merkle Tree
    
    Mirrors Rust: merkle crate pattern
    """
    if not items:
        return sha256_hash(b'')
    
    if len(items) == 1:
        return items[0]
    
    # Pad to even number
    if len(items) % 2 == 1:
        items = items + [items[-1]]
    
    # Build tree bottom-up
    while len(items) > 1:
        next_level = []
        for i in range(0, len(items), 2):
            combined = items[i] + items[i + 1]
            next_level.append(sha256_hash(combined))
        items = next_level
    
    return items[0]


def hash_object(obj: Any) -> bytes:
    """Hash a Python object (serializes to JSON first)"""
    data = json.dumps(obj, sort_keys=True, default=str).encode()
    return sha256_hash(data)


def hash_hex(obj: Any) -> str:
    """Hash object and return hex string"""
    return hash_object(obj).hex()

File: rust-core/jasprchain/src/crypto/hash.rs
//! Hash functions

use sha2::{Sha256, Digest};

/// SHA-256 hash
pub fn sha256(data: &[u8]) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(data);
    hasher.finalize().into()
}

/// SHA-256 hash as hex string
pub fn sha256_hex(data: &[u8]) -> String {
    hex::encode(sha256(data))
}

/// Calculate Merkle root from list of hashes
pub fn merkle_root(items: &[[u8; 32]]) -> [u8; 32] {
    if items.is_empty() {
        return sha256(b"");
    }
    
    if items.len() == 1 {
        return items[0];
    }
    
    let mut current: Vec<[u8; 32]> = items.to_vec();
    
    // Pad to even number
    if current.len() % 2 == 1 {
        current.push(*current.last().unwrap());
    }
    
    while current.len() > 1 {
        let mut next = Vec::new();
        for chunk in current.chunks(2) {
            let mut combined = Vec::new();
            combined.extend_from_slice(&chunk[0]);
            combined.extend_from_slice(&chunk[1]);
            next.push(sha256(&combined));
        }
        current = next;
        
        if current.len() > 1 && current.len() % 2 == 1 {
            current.push(*current.last().unwrap());
        }
    }
    
    current[0]
}
